### ISSUE TITLE
Better logging of cached partial renders

### ISSUE THREAD

**Issue by dhh** [rails/rails#23879]
Now that we have the lovely [X/Y cache hits] formats for partial collections, I think we should expand that to single partial renders as well.

Here's what things currently look like:

```
Read fragment views/v1/2914079/v1/2914079/recordings/70182313-20160225015037000000/d0bdf2974e1ef6d31685c3b392ad0b74 (0.6ms)
Rendered messages/_message.html.erb in 1.2 ms
Write fragment views/v1/2914079/v1/2914079/recordings/70182313-20160225015037000000/3b4e249ac9d168c617e32e84b99218b5 (1.1ms)
Rendered recordings/threads/_thread.html.erb in 1.5 ms
```

I'd like to see this instead:

```
Rendered messages/_message.html.erb in 1.2 ms [cache hit]
Rendered recordings/threads/_thread.html.erb in 1.5 ms [cache miss]
```

**Comment by st0012** [comment]
I'm working on this :smile:

**Comment by kaspth** [comment]
@st0012 Neat! But mind the milestone, I won't have much time to review a pull request before Rails 5.0 is out.

**Comment by st0012** [comment]
@kaspth ok, but I will still mention you when I done this.

**Comment by kaspth** [comment]
@st0012 by all means do! I'm just setting expectations 😁

**Comment by st0012** [comment]
@kaspth Hi, I noticed that collection partials' caching functionality and instrument option was implemented [here](https://github.com/rails/rails/blob/b4700de1ce21599b500d43d8138184ee7ae81407/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb#L17), but partial rendering doesn't have something like that. So does this issue also require implementing partial caching mechanism?

**Comment by kaspth** [comment]
No, there's already `cache` in views for that. We need to make that able to fill in `:cache_hit` in the instrumentation payload like we do for collection caching 😁

**Comment by st0012** [comment]
So that means we need to shift the instrumenting to earlier actions, like `cache` in `caching_helper`?

**Comment by st0012** [comment]
Oh, `fragment_for` might be a better [place](https://github.com/rails/rails/blob/b4700de1ce21599b500d43d8138184ee7ae81407/actionview%2Flib%2Faction_view%2Fhelpers%2Fcache_helper.rb#L207)

**Comment by kaspth** [comment]
@st0012 try researching it a bit more and write something more substantial. This isn't a chat app 😁

**Comment by st0012** [comment]
@kaspth Sorry. I already started working on that today, I just curious about what approach to take (and it seems that I misunderstood something). I hope I can open a PR before the end of this week.

**Comment by kaspth** [comment]
No worries, and don't worry about getting the PR ready for the end of the week. We can't merge this before the Rails 5 stable branch is created.

**Comment by st0012** [comment]
After some research, here's what I found:
The reason collection partials caching can be instrumented is because the `cached` option would be passed in `render` method, like
```
<%= render partial: 'notifications/notification', collection: @notifications, cached: true %>
```
And the collection partial's caching is implemented in the partial rendering level. So the `Logsubscriber` knows that whether the collections should be cached and whether cache is success. The process may looks like this:
```
PartialRenderer#render_collection
PartialRenderer::CollectionCaching#cache_collection_render (pass payload here)
(Caching collections and put result into payload)
Done
```

However, normal partial caching happens in the last step of view rendering, which is process the template content. That means we can't know if the view uses cache nor it's cache result, unless we figure out some way to pass the payload deep into that level, but that will touch a lot of things I think.

**Comment by kaspth** [comment]
That wasn't really what I meant by research. This is what you should look into:

> That means we can't know if the view uses cache nor it's cache result, unless we figure out some way to pass the payload deep into that level, but that will touch a lot of things I think.

What we can take from the collection caching stuff is how it fills in the instrumentation payload as it's running, that's what we need to do for normal partial renders too 😁

**Comment by kaspth** [comment]
@st0012 hey, how far did you come with this? I'm curious to see what you got 😁

**Comment by st0012** [comment]
@kaspth Hi, sorry that I just forgot this issue for last few months. I just restudied it after Rails 5 officially came out. 😓
I think I know what to do with this issue after restudied the codebase and our conversation, and I will do some naive implementation this weekend and attach a PR this weekend.

**Comment by st0012** [comment]
@kaspth I just attached my PR, hope that you can take a look and give me some advice 😁

**Comment by st0012** [comment]
@kaspth I thinks this can be closed?

**Comment by kaspth** [comment]
@st0012 right you are! I'm too used to pull requests having "Fixes issue-number" and GitHub closing issues automatically that I forget it's not always so. That's a tip for your next pull request and thanks 😁

Fixed in #25825.
